### ISSUE TITLE
Detect file moves and don't reupload

### ISSUE THREAD

**Issue by DarkArc** [duplicati/duplicati#4432]
- [X] I have searched open and closed issues for duplicates.

----------------------------------------

## Environment info

- **Duplicati version**: 2.0.5.1_beta_2020-01-18
- **Operating system**: Fedora 33
- **Backend**: Dropbox

## Description

This is a feature request/enhancement request. Duplicati when creating an incremental backup, should look for file moves (possibly with a file size threshold). Currently moving several large files from one directory to another, both within folders backed up, result in the file being reuploaded, eating a ton of bandwidth.

It would be ideal Duplicati could detect this and somehow record "it's this object, but it's not stored there anymore, it's stored here" and avoid the reupload. This would be conceptually very similar to how git can detect moves for files that have small amounts of/no changes.

This request is based on an observation of very long backups (when files have been moved, but not modified) vs very short backups (when nothing has changed at all).

----

P.S. Thanks for this awesome project :)

**Comment by DarkArc** [comment]
Hm... Looking back at the backup I just ran that follows this pattern, I had a large spike in the progressbar. Maybe it's not necessarily reuploading everything, but something else is going on?

The only reason I say/think it's related to reuploading is that my upload bandwidth remains pegged throughout the backup.

**Comment by drwtsn32x** [comment]
This shouldn't be the case.  Moving data around may cause it to be re-analyzed as if it were new or changed files, but ultimately if the blocks that comprise the data already exist on the back end, then the data won't need to be re-uploaded.

**Comment by DarkArc** [comment]
Perhaps this is a bug then? Possibly with the dropbox backend?

**Comment by drwtsn32x** [comment]
I wouldn't think so.... all files Duplicati backs up are packaged and stored in a single folder on the back end, so moving them around on the source side doesn't matter.

**Comment by ts678** [comment]
> observation of very long backups vs very short backups when the same files are untouched but moved.

I'm unclear on what this says. Looking right of "vs" says moving files but not otherwise touching is fast. What's to left of "vs."?
If touched files are moved, the question is how they changed when touched. When files change, they're no longer the "same".

You can certainly get an idea if file changed due to move by comparing before and after hashes from `md5sum`, `sha1sum`, etc.

**Comment by DarkArc** [comment]
> > observation of very long backups vs very short backups when the same files are untouched but moved.
> 
> I'm unclear on what this says. Looking right of "vs" says moving files but not otherwise touching is fast. What's to left of "vs."?
> If touched files are moved, the question is how they changed when touched. When files change, they're no longer the "same".

Yeah, I see what you're saying. I was tired and in a hurry, my bad. I've clarified this in the issue description.

----

I took a closer look at this tonight. Just renaming a directory with 1,218 files and 269 subfolders resulted in an increase of 200MB -- it's possible that was influence by some other changes to config files in this backup configuration, but at 200MB I'd be highly surprised if that's the case.

This is further supported by just rerunning the backup taking 70 MB, still not great but a far cry from 200MB.

Additionally, comparing the two, just checking the files to get to the point where the backup is uploaded took significantly longer in the case where a parent directory was renamed. (e.g. `Music/Files/Source` -> `Music/Files/Source - 2`; `Music` being the top level directory that's backed up in this directory tree). 

Is there some kind of bottleneck in the examination of existing files perhaps? Single threaded hashing of files as the directory tree is walked? etc?

**Comment by ts678** [comment]
> Additionally, comparing the two, just checking the files to get to the point where the backup is uploaded took significantly longer in the case where a parent directory was renamed

This one, at least, is explainable because the files that are moved are seen as "new" (path didn't exist before) so are read through. Contents will be able to reuse the blocks of the file when at previous location, but all of this analysis adds to the processing time.
The job log in its upper right corner on `Source files` will probably show the usual total count in "Examined", with big numbers showing for `Deleted`, `Added`, and `Opened` because a file vanished, a new one showed up, and it was opened to read contents.
Anything in `Modified` should be a red flag that something might be changing the files. Is the area completely idle, or updating?

![image](https://user-images.githubusercontent.com/40771998/107311920-4bbb4900-6a5d-11eb-913f-48ccad5f2925.png)

For a test, I made a folder to backup. In a subfolder I added a set of files, backed up, ran again, then renamed subfolder and ran.
Looking in `Complete log` of three job logs, in sequence, saw "BytesUploaded" of 482369993, 0, and 4266. The 0 is because an unchanged backup by default doesn't even create a backup version. If rerunning the backup uploaded, something had changed.

I tested on Windows, but same result should hold for Linux. Sometimes moves can change non-content things like timestamps.
If you want to see why each file got handled (and I suggest a small test with a few well-controlled files) look at verbose logging.
About --> Show log --> Live is a good start, but log file is also possible if live log is too unwieldy. You see why a file is handled:

> 2020-09-29 16:24:34 -04 - [Verbose-Duplicati.Library.Main.Operation.Backup.FilePreFilterProcess.FileEntry-CheckFileForChanges]: Checking file for changes C:\PortableApps\Notepad++Portable\App\Notepad++\backup\webpages.txt@2020-09-25_201932, new: False, timestamp changed: True, size changed: True, metadatachanged: True, 9/29/2020 8:01:27 PM vs 9/29/2020 7:10:55 PM
